prov/efa/src/efa_prov_info.c

```diff
@@ -259,118 +439,55 @@ static int efa_alloc_fid_nic(struct fi_info *fi, struct efa_device *device)
+void efa_prov_info_set_hmem_flags(struct fi_info *prov_info)
+{
+	if (prov_info->ep_attr->type == FI_EP_RDM &&
 	    (ofi_hmem_is_initialized(FI_HMEM_CUDA) ||
 	     ofi_hmem_is_initialized(FI_HMEM_NEURON))) {
-		info->caps			|= FI_HMEM;
-		info->tx_attr->caps		|= FI_HMEM;
-		info->rx_attr->caps		|= FI_HMEM;
-		info->domain_attr->mr_mode	|= FI_MR_HMEM;
-	}
-#endif
-
-	EFA_DBG(FI_LOG_DOMAIN, "Domain attribute :\n"
-				"\t info->domain_attr->cq_cnt		= %zu\n"
-				"\t info->domain_attr->ep_cnt		= %zu\n"
-				"\t info->domain_attr->rx_ctx_cnt	= %zu\n"
-				"\t info->domain_attr->tx_ctx_cnt	= %zu\n"
-				"\t info->domain_attr->max_ep_tx_ctx	= %zu\n"
-				"\t info->domain_attr->max_ep_rx_ctx	= %zu\n",
-				info->domain_attr->cq_cnt,
-				info->domain_attr->ep_cnt,
-				info->domain_attr->tx_ctx_cnt,
-				info->domain_attr->rx_ctx_cnt,
-				info->domain_attr->max_ep_tx_ctx,
-				info->domain_attr->max_ep_rx_ctx);
-
-	info->tx_attr->iov_limit = device->efa_attr.max_sq_sge;
-	info->tx_attr->size = rounddown_power_of_two(device->efa_attr.max_sq_wr);
-	if (info->ep_attr->type == FI_EP_RDM) {
-		info->tx_attr->inject_size = device->efa_attr.inline_buf_size;
-	} else if (info->ep_attr->type == FI_EP_DGRAM) {
-                /*
-                 * Currently, there is no mechanism for EFA layer (lower layer)
-                 * to discard completions internally and FI_INJECT is not optional,
-                 * it can only be disabled by setting inject_size to 0. RXR
-                 * layer does not have this issue as completions can be read from
-                 * the EFA layer and discarded in the RXR layer. For dgram
-                 * endpoint, inject size needs to be set to 0
-                 */
-		info->tx_attr->inject_size = 0;
-	}
-	info->rx_attr->iov_limit = device->efa_attr.max_rq_sge;
-	info->rx_attr->size = rounddown_power_of_two(device->efa_attr.max_rq_wr / info->rx_attr->iov_limit);
-
-	EFA_DBG(FI_LOG_DOMAIN, "Tx/Rx attribute :\n"
-				"\t info->tx_attr->iov_limit		= %zu\n"
-				"\t info->tx_attr->size			= %zu\n"
-				"\t info->tx_attr->inject_size		= %zu\n"
-				"\t info->rx_attr->iov_limit		= %zu\n"
-				"\t info->rx_attr->size			= %zu\n",
-				info->tx_attr->iov_limit,
-				info->tx_attr->size,
-				info->tx_attr->inject_size,
-				info->rx_attr->iov_limit,
-				info->rx_attr->size);
-
-
-	info->ep_attr->max_msg_size		= device->ibv_port_attr.max_msg_sz;
-	info->ep_attr->max_order_raw_size	= device->ibv_port_attr.max_msg_sz;
-	info->ep_attr->max_order_waw_size	= device->ibv_port_attr.max_msg_sz;
-
-	/* Set fid nic attributes. */
-	ret = efa_alloc_fid_nic(info, device);
-	if (ret) {
-		EFA_WARN(FI_LOG_FABRIC,
-			 "Unable to allocate fid_nic: %s\n", fi_strerror(-ret));
-		return ret;
+		prov_info->caps			|= FI_HMEM;
+		prov_info->tx_attr->caps		|= FI_HMEM;
+		prov_info->rx_attr->caps		|= FI_HMEM;
+		prov_info->domain_attr->mr_mode	|= FI_MR_HMEM;
 	}
-
-	return 0;
 }
+#else
+void efa_prov_info_set_hmem_flags(struct fi_info *prov_info)
```

Comment:
Why not have the `#if` inside the function?

Reply:
`#if` outside function is libfabric's preference.